lib/scala/testkit/src/main/java/org/enso/testkit/RetryTestRule.java

```diff
@@ -0,0 +1,56 @@
+package org.enso.testkit;
+
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+/**
+ * Flaky test specification for JUnit.
+ *
+ * <p>Inspired by <a href="https://stackoverflow.com/a/8301639/4816269">this SO answer</a>.
+ *
+ * <p>Use it like this:
+ *
+ * <pre>
+ * public class MyTest {
+ *   &#64;Rule
+ *   public RetryTestRule retry = new RetryTestRule(3);
+ *   &#64;Test
+ *   public void myTest() {...}
+ * }
+ * </pre>
+ */
+public class RetryTestRule implements TestRule {
+  private int retryCount;
+
+  public RetryTestRule(int retryCount) {
+    this.retryCount = retryCount;
+  }
+
+  @Override
+  public Statement apply(Statement base, Description description) {
+    return statement(base, description);
+  }
+
+  private Statement statement(final Statement base, final Description description) {
+    return new Statement() {
+      @Override
+      public void evaluate() throws Throwable {
+        Throwable caughtThrowable = null;
+
+        for (int i = 0; i < retryCount; i++) {
+          try {
+            base.evaluate();
+            return;
+          } catch (Throwable t) {
+            caughtThrowable = t;
+            System.err.println(description.getDisplayName() + ": run " + (i + 1) + " failed");
+          }
+        }
+        System.err.println(
+            description.getDisplayName() + ": giving up after " + retryCount + " failures");
+        throw caughtThrowable;
```

Comment:
Wouldn't it be better to attach this description to the a throwable that gets `throw`?
```java
Set<Throwable> caughtThrowables = new TreeSet<>(compareThrowableMessageAndStack);
...
var err = new AssertionError(description.getDisplayName() + ": giving up after " + retryCount + " failures");
for (var t : caughtThrowables) {
  var n = err; while (n.getCause() != null) n = n.getCause();
  n.initCause(t);
}
throw err;
```
e.g. make sure all the information would get into `junit-report.xml` rather than be dispersed in the stderr?

Reply:
Yes, that definitelly seems like a good idea. Implemented in d437a0615772cb1cd6f357bb6aff4f5df022081b. Now, if there is an error in a test with `RetryTestRule`, the stdoutput looks like this:
```
sbt:downloader> tst *HttpDownloaderTest
[info] compiling 1 Java source to /home/pavel/dev/enso/lib/scala/testkit/target/scala-2.13/classes ...
Serving files from directory /home/pavel/dev/enso/tools/http-test-helper/www-files
HTTP server started at /127.0.0.1:8090
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
org.enso.downloader.http.HttpDownloaderTest.fetchStringTest(org.enso.downloader.http.HttpDownloaderTest): run 1 failed
org.enso.downloader.http.HttpDownloaderTest.fetchStringTest(org.enso.downloader.http.HttpDownloaderTest): run 2 failed
org.enso.downloader.http.HttpDownloaderTest.fetchStringTest(org.enso.downloader.http.HttpDownloaderTest): run 3 failed
fetchStringTest(org.enso.downloader.http.HttpDownloaderTest): giving up after 3 failures
[error] Test org.enso.downloader.http.HttpDownloaderTest.fetchStringTest failed: java.lang.AssertionError: fetchStringTest(org.enso.downloader.http.HttpDownloaderTest): giving up after 3 failures, took 0.342 sec
[error]     at org.enso.testkit.RetryTestRule$1.evaluate(RetryTestRule.java:61)
[error]     ...
[error] Caused by: java.lang.AssertionError: 
[error] Expected: a string containing "Hello XXX"
[error]      but: was "Hello, world!"
[error]     at org.enso.downloader.http.HttpDownloaderTest.fetchStringTest(HttpDownloaderTest.java:69)
[error]     at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[error]     at java.lang.reflect.Method.invoke(Method.java:580)
[error]     ...
[error] Caused by: java.lang.AssertionError: 
[error] Expected: a string containing "Hello XXX"
[error]      but: was "Hello, world!"
[error]     ... 41 more
[error] Caused by: java.lang.AssertionError: 
[error] Expected: a string containing "Hello XXX"
[error]      but: was "Hello, world!"
[error]     ... 41 more
Finalizing HTTP server...
[error] Failed: Total 6, Failed 1, Errors 0, Passed 5
[error] Failed tests:
[error] 	org.enso.downloader.http.HttpDownloaderTest
[error] (Test / testOnly) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 4 s, completed Mar 11, 2024, 12:16:58 PM
```

And the generated test-report in `enso/lib/scala/downloader/target/test-reports/TEST-org.enso.downloader.http.HttpDownloaderTest.xml` has the following content:
```xml
<?xml version='1.0' encoding='UTF-8'?>
<testsuite hostname="pavel-thinkpad" name="org.enso.downloader.http.HttpDownloaderTest" tests="6" errors="0" failures="1" skipped="0" time="0.435" timestamp="2024-03-11T12:16:57">
          <properties>
      <property name="java.specification.version" value="21"/><property name="sun.jnu.encoding" value="UTF-8"/><property name="sun.arch.data.model" value="64"/><property name="java.vendor.url" value="https://www.graalvm.org/"/><property name="sun.boot.library.path" value="/home/pavel/.sdkman/candidates/java/21.0.2-graalce/lib"/><property name="sun.java.command" value="/home/pavel/.sdkman/candidates/sbt/1.9.7/bin/sbt-launch.jar"/><property name="jdk.debug" value="release"/><property name="java.specification.vendor" value="Oracle Corporation"/><property name="java.version.date" value="2024-01-16"/><property name="java.home" value="/home/pavel/.sdkman/candidates/java/21.0.2-graalce"/><property name="file.separator" value="/"/><property name="java.vm.compressedOopsMode" value="Zero based"/><property name="line.separator" value="
"/><property name="java.vm.specification.vendor" value="Oracle Corporation"/><property name="java.specification.name" value="Java Platform API Specification"/><property name="sun.management.compiler" value="HotSpot 64-Bit Tiered Compilers"/><property name="java.runtime.version" value="21.0.2+13-jvmci-23.1-b30"/><property name="user.name" value="pavel"/><property name="log4j.ignoreTCL" value="true"/><property name="file.encoding" value="UTF-8"/><property name="jnidispatch.path" value="/home/pavel/.cache/JNA/temp/jna4066371523924934014.tmp"/><property name="java.vendor.version" value="GraalVM CE 21.0.2+13.1"/><property name="jna.loaded" value="true"/><property name="java.io.tmpdir" value="/tmp"/><property name="java.version" value="21.0.2"/><property name="java.vm.specification.name" value="Java Virtual Machine Specification"/><property name="native.encoding" value="UTF-8"/><property name="java.library.path" value="/usr/java/packages/lib:/usr/lib64:/lib64:/lib:/usr/lib"/><property name="stderr.encoding" value="UTF-8"/><property name="java.vendor" value="GraalVM Community"/><property name="scala.ext.dirs" value="/home/pavel/.sbt/1.0/java9-rt-ext-graalvm_community_21_0_2"/><property name="sun.io.unicode.encoding" value="UnicodeLittle"/><property name="java.class.path" value="/home/pavel/.sdkman/candidates/sbt/1.9.7/bin/sbt-launch.jar"/><property name="java.vm.vendor" value="GraalVM Community"/><property name="jline.shutdownhook" value="false"/><property name="user.timezone" value="Europe/Prague"/><property name="java.vm.specification.version" value="21"/><property name="os.name" value="Linux"/><property name="sun.java.launcher" value="SUN_STANDARD"/><property name="user.country" value="US"/><property name="sun.cpu.endian" value="little"/><property name="user.home" value="/home/pavel"/><property name="user.language" value="en"/><property name="sbt.script" value="/home/pavel/.sdkman/candidates/sbt/current/bin/sbt"/><property name="sbt.ipcsocket.tmpdir" value="/run/user/1000/.sbtea728646/ipcsocket"/><property name="swoval.tmpdir" value="/run/user/1000/.sbtea728646/swoval"/><property name="jdk.internal.vm.ci.enabled" value="true"/><property name="jline.esc.timeout" value="0"/><property name="stdout.encoding" value="UTF-8"/><property name="path.separator" value=":"/><property name="os.version" value="6.5.0-21-generic"/><property name="jna.nosys" value="true"/><property name="java.runtime.name" value="OpenJDK Runtime Environment"/><property name="java.vm.name" value="OpenJDK 64-Bit Server VM"/><property name="jna.platform.library.path" value="/usr/lib/x86_64-linux-gnu:/lib/x86_64-linux-gnu:/usr/lib64:/lib64:/usr/lib:/lib:/lib32:/usr/lib/x86_64-linux-gnu/libfakeroot"/><property name="java.vendor.url.bug" value="https://github.com/oracle/graal/issues"/><property name="user.dir" value="/home/pavel/dev/enso"/><property name="os.arch" value="amd64"/><property name="java.vm.info" value="mixed mode, sharing"/><property name="java.vm.version" value="21.0.2+13-jvmci-23.1-b30"/><property name="java.class.version" value="65.0"/>
    </properties>
          <testcase classname="org.enso.downloader.http.HttpDownloaderTest" name="fetchStringTest" time="0.342">
                      <failure message="java.lang.AssertionError: fetchStringTest(org.enso.downloader.http.HttpDownloaderTest): giving up after 3 failures" type="sbt.ForkMain$ForkError">sbt.ForkMain$ForkError: java.lang.AssertionError: fetchStringTest(org.enso.downloader.http.HttpDownloaderTest): giving up after 3 failures
	at org.enso.testkit.RetryTestRule$1.evaluate(RetryTestRule.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at com.novocode.junit.JUnitTask.execute(JUnitTask.java:64)
	at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:414)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: sbt.ForkMain$ForkError: java.lang.AssertionError: 
Expected: a string containing &quot;Hello XXX&quot;
     but: was &quot;Hello, world!&quot;
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
	at org.enso.downloader.http.HttpDownloaderTest.fetchStringTest(HttpDownloaderTest.java:69)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.enso.testkit.RetryTestRule$1.evaluate(RetryTestRule.java:46)
	... 31 more
Caused by: sbt.ForkMain$ForkError: java.lang.AssertionError: 
Expected: a string containing &quot;Hello XXX&quot;
     but: was &quot;Hello, world!&quot;
	... 41 more
Caused by: sbt.ForkMain$ForkError: java.lang.AssertionError: 
Expected: a string containing &quot;Hello XXX&quot;
     but: was &quot;Hello, world!&quot;
	... 41 more
</failure>
                    </testcase><testcase classname="org.enso.downloader.http.HttpDownloaderTest" name="handleResourceNotFound" time="0.008">
                      
                    </testcase><testcase classname="org.enso.downloader.http.HttpDownloaderTest" name="downloadSmallFileTest" time="0.015">
                      
                    </testcase><testcase classname="org.enso.downloader.http.HttpDownloaderTest" name="handleRedirectTest" time="0.051">
                      
                    </testcase><testcase classname="org.enso.downloader.http.HttpDownloaderTest" name="downloadBigFileWithProgress" time="0.015">
                      
                    </testcase><testcase classname="org.enso.downloader.http.HttpDownloaderTest" name="fetchStringWithProgress" time="0.004">
                      
                    </testcase>
          <system-out><![CDATA[]]></system-out>
          <system-err><![CDATA[]]></system-err>
        </testsuite>
```